File: apps/web/lib/access.ts
/**
 * Utility functions for checking subscription-based feature access
 */

export interface UserSubscription {
  status: string;
  expiry_date: string;
  plan_details: {
    features: string[];
  };
}

/**
 * Checks if the user has access to a specific premium feature
 * @param subscription The user's active subscription object
 * @param featureKey The key of the feature to check (e.g., 'ai_tutor', 'premium_materials')
 * @returns boolean True if user has access, false otherwise
 */
export function hasFeature(subscription: UserSubscription | null | undefined, featureKey: string): boolean {
  if (!subscription) return false;
  
  if (subscription.status !== 'ACTIVE') return false;
  
  const expiryDate = new Date(subscription.expiry_date);
  const now = new Date();
  
  if (expiryDate <= now) return false;
  
  // If the plan has features array, check if it includes the requested feature
  if (subscription.plan_details?.features) {
    // If the plan has '*' feature, it unlocks everything
    if (subscription.plan_details.features.includes('*')) return true;
    
    return subscription.plan_details.features.includes(featureKey);
  }
  
  return false;
}

/**
 * Common feature keys used in the system
 */
export const FEATURES = {
  AI_TUTOR: 'ai_tutor',
  PREMIUM_MATERIALS: 'premium_materials',
  ADVANCED_MOCK_EXAM: 'advanced_mock_exam',
  ANALYTICS: 'analytics',
};

File: apps/api/subscriptions/access.py
"""Server-side mirror of apps/web/lib/access.ts - the single source of truth
for "does this student have an active package" and "does their plan include
this feature". Nothing else in the codebase should re-implement this check;
import from here.
"""
from django.utils import timezone

def has_admin_granted_access(user):
    from support.models import StudentProfile
    try:
        profile = user.student_profile
        if profile.access_origin == 'ADMIN_GRANTED':
            if profile.admin_access_expiry is None or profile.admin_access_expiry > timezone.now():
                return True
    except StudentProfile.DoesNotExist:
        pass
    return False


def get_active_subscription(user):
    """Returns the student's current active, non-expired Subscription, or
    None. 'Active' means status=='ACTIVE' AND expiry_date is in the future -
    matching Subscription.is_active exactly, but as a queryset so callers
    that need the object (not just a bool) don't have to iterate."""
    from .models import Subscription

    return (
        Subscription.objects.filter(
            student=user, status='ACTIVE', expiry_date__gt=timezone.now()
        )
        .select_related('plan')
        .order_by('-expiry_date')
        .first()
    )


def has_active_subscription(user):
    return get_active_subscription(user) is not None


def has_feature(user, feature_key):
    """Mirrors lib/access.ts hasFeature(): '*' in plan.features unlocks
    everything, otherwise the exact key must be present."""
    if has_admin_granted_access(user):
        return True

    subscription = get_active_subscription(user)
    if subscription is None:
        return False

    features = subscription.plan.features or []
    if '*' in features:
        return True
    return feature_key in features
